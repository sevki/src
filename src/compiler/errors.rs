use std::ops::Range;

use crate::lexer::Coord;

pub struct Errors<'a>(Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>);

impl<'a> From<Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>>
    for Errors<'a>
{
    fn from(
        value: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'a>, &'a str>>,
    ) -> Self {
        Self(value)
    }
}

impl<'a> IntoIterator for Errors<'a> {
    type Item = Range<Coord>;

    type IntoIter = <Vec<std::ops::Range<Coord>> as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.0
            .into_iter()
            .map(|error| match error.error {
                lalrpop_util::ParseError::InvalidToken { location } => location..location,
                lalrpop_util::ParseError::UnrecognizedEof {
                    location,
                    expected: _,
                } => location..location,
                lalrpop_util::ParseError::UnrecognizedToken { token, expected: _ } => {
                    token.0..token.2
                }
                lalrpop_util::ParseError::ExtraToken { token } => token.0..token.2,
                lalrpop_util::ParseError::User { error: _ } => todo!(),
            })
            .collect::<Vec<_>>()
            .into_iter()
    }
}

fn handle_errors(
    errors: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token, &str>>,
    src: &str,
) -> String {
    let mut pretty = String::new();
    let mut last_end = Coord::default();

    for error in errors {
        match error.error {
            lalrpop_util::ParseError::InvalidToken { location: _ } => todo!(),
            lalrpop_util::ParseError::UnrecognizedEof {
                location: _,
                expected: _,
            } => todo!(),
            lalrpop_util::ParseError::UnrecognizedToken { token, expected } => {
                // find the line and column of the start and end tokens,
                // and print the line with a caret pointing to the error
                let start = token.0;
                let end = token.2;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                // pretty.push_str(&src[last_end..start]);
                pretty.push_str(&format!(
                    "error: unexpected token {:?}, expected one of {:?}\n",
                    token.1, expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
                last_end = end;
            }
            lalrpop_util::ParseError::ExtraToken { token: _ } => todo!(),
            lalrpop_util::ParseError::User { error: _ } => todo!(),
        };
    }
    // pretty.push_str(&src[last_end..]);
    pretty
}
