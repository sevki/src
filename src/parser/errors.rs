use crate::lexer::Coord;

pub fn pretty_errors<'input>(
    src: &'input str,
    errors: Vec<lalrpop_util::ErrorRecovery<Coord, crate::lexer::Token<'_>, &str>>,
) -> String {
    let mut pretty = String::new();

    for error in errors {
        match error.error {
            lalrpop_util::ParseError::InvalidToken { location: _ } => todo!(),
            lalrpop_util::ParseError::UnrecognizedEof { location, expected } => {
                let start = location;
                let end = location;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                pretty.push_str(&format!(
                    "error: unexpected end of file, expected one of {:?}\n",
                    expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
            }
            lalrpop_util::ParseError::UnrecognizedToken { token, expected } => {
                // find the line and column of the start and end tokens,
                // and print the line with a caret pointing to the error
                let start = token.0;
                let end = token.2;
                let start_line = start.line;
                let end_line = end.line;
                let line = &src[start_line..end_line];
                let start_col = start.col;
                let end_col = end.col;
                // pretty.push_str(&src[last_end..start]);
                // expected is double quoted, so we need to remove the quotes
                // which ends up printing like   "\"(\""
                let expected: Vec<&str> = expected.iter().map(|x| &x[1..x.len() - 1]).collect();
                pretty.push_str(&format!(
                    "error: unexpected token {:?}, expected one of {:?}\n",
                    token.1, expected
                ));
                pretty.push_str(&line);
                pretty.push_str("\n");
                pretty.push_str(&" ".repeat(start_col));
                pretty.push_str(&"^".repeat(end_col - start_col));
            }
            lalrpop_util::ParseError::ExtraToken { token: _ } => todo!(),
            lalrpop_util::ParseError::User { error: _ } => todo!(),
        };
    }
    // pretty.push_str(&src[last_end..]);
    pretty
}
