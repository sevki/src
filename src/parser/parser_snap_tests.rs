
use crate::parser::errors::pretty_errors;
use insta::assert_snapshot;


#[cfg(test)]
#[okstd::test]
fn test_empty_parser() {
    let input = " ";
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()), @r###"
    Module(
        [],
    )
    "###);
}

#[okstd::test]
fn test_fn_call_parser_with_multiple_args_and_strings() {
    let input = "fn some()[] {let a = some_fnExpr(1, \"2\", 3)}";
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    if !errors.is_empty() {
        panic!("{}", pretty_errors(&input, errors));
    }
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_fn_def_parser() {
    let input = r#"fn call(a:b, b:c) [throws, awaits, execs] {
        call(1+1)
        let a = 1
    } when throws {
        raise(1)
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_effect() {
    let input = r#"effect VM: async + throws + execs {
        catch() []
        await<T>(f: Future<T>) [] -> T
        exec(arg0: string, args: stringvec) []
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_struct_parser() {
    let input = r#"struct VM {
        a: string
        b: string
    }"#;
    let mut errors = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    assert!(errors.is_empty());
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}

#[okstd::test]
fn test_enum_parser() {
    let input = r#"use { exec } from host

    effect Make: async + throws + execs + reads + writes {
        catch() [throws]
        await<T>(f: Future<T>) [async, throws] -> T
        exec(arg0: string, args: stringvec) [Make] -> i32
        read(name: string) [reads] -> string
        write(name: string, value: string) [writes]
    }
    
    struct Local {
        host: host
    }
    
    impl Make for Local {
        fn catch(self) [throws] {
        } 
        fn await<T>(f: Future<T>) [async, trhows] -> T {
            yield()
        }
        fn exec(self, arg0: string, args: vec<string>) [Vm] -> i32 {
            self.host.read("jobserver")
            if self.host.exec(arg0, args) {
                raise(1)
            }
        }
    }"#;
    let mut errors: Vec<lalrpop_util::ErrorRecovery<usize, crate::lexer::Token, &str>> = vec![];
    let wrapper = crate::lexer::TripleIterator::new(input);
    let t = crate::parser::src::SourceParser::new().parse(&mut errors, wrapper);
    if !errors.is_empty() {
        panic!("{}", pretty_errors(&input, errors));
    }
    assert_snapshot!(format!("{:#?}", t.unwrap()));
}
