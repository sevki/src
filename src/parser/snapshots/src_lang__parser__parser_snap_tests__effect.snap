---
source: src/parser/parser_snap_tests.rs
expression: "format!(\"{:#?}\", t.unwrap())"
---
Module(
    [
        Spanned(
            Location {
                offset: 0,
                line: 0,
                col: 0,
            },
            EffectDef(
                EffectDef(
                    Spanned(
                        Location {
                            offset: 0,
                            line: 0,
                            col: 0,
                        },
                        KeywordAndVisibility(
                            Spanned(
                                Location {
                                    offset: 0,
                                    line: 0,
                                    col: 0,
                                },
                                Effect,
                                Location {
                                    offset: 6,
                                    line: 0,
                                    col: 6,
                                },
                            ),
                            Spanned(
                                Location {
                                    offset: 0,
                                    line: 0,
                                    col: 0,
                                },
                                Private,
                                Location {
                                    offset: 0,
                                    line: 0,
                                    col: 0,
                                },
                            ),
                        ),
                        Location {
                            offset: 6,
                            line: 0,
                            col: 6,
                        },
                    ),
                    Spanned(
                        Location {
                            offset: 7,
                            line: 0,
                            col: 7,
                        },
                        Ident(
                            "VM",
                            None,
                        ),
                        Location {
                            offset: 9,
                            line: 0,
                            col: 9,
                        },
                    ),
                    [
                        Spanned(
                            Location {
                                offset: 11,
                                line: 0,
                                col: 11,
                            },
                            Ident(
                                "async",
                                None,
                            ),
                            Location {
                                offset: 16,
                                line: 0,
                                col: 16,
                            },
                        ),
                        Spanned(
                            Location {
                                offset: 19,
                                line: 0,
                                col: 19,
                            },
                            Ident(
                                "throws",
                                None,
                            ),
                            Location {
                                offset: 25,
                                line: 0,
                                col: 25,
                            },
                        ),
                        Spanned(
                            Location {
                                offset: 28,
                                line: 0,
                                col: 28,
                            },
                            Ident(
                                "execs",
                                None,
                            ),
                            Location {
                                offset: 33,
                                line: 0,
                                col: 33,
                            },
                        ),
                    ],
                    Block(
                        [
                            Spanned(
                                Location {
                                    offset: 40,
                                    line: 1,
                                    col: 4,
                                },
                                Prototype {
                                    name: Spanned(
                                        Location {
                                            offset: 40,
                                            line: 1,
                                            col: 4,
                                        },
                                        Ident(
                                            "catch",
                                            None,
                                        ),
                                        Location {
                                            offset: 45,
                                            line: 1,
                                            col: 9,
                                        },
                                    ),
                                    args: [],
                                    ret: None,
                                    effects: [],
                                },
                                Location {
                                    offset: 50,
                                    line: 1,
                                    col: 14,
                                },
                            ),
                            Spanned(
                                Location {
                                    offset: 55,
                                    line: 2,
                                    col: 4,
                                },
                                Prototype {
                                    name: Spanned(
                                        Location {
                                            offset: 55,
                                            line: 2,
                                            col: 4,
                                        },
                                        Ident(
                                            "await",
                                            Some(
                                                [
                                                    Spanned(
                                                        Location {
                                                            offset: 61,
                                                            line: 2,
                                                            col: 10,
                                                        },
                                                        Ident(
                                                            "T",
                                                            None,
                                                        ),
                                                        Location {
                                                            offset: 62,
                                                            line: 2,
                                                            col: 11,
                                                        },
                                                    ),
                                                ],
                                            ),
                                        ),
                                        Location {
                                            offset: 63,
                                            line: 2,
                                            col: 12,
                                        },
                                    ),
                                    args: [
                                        Spanned(
                                            Location {
                                                offset: 64,
                                                line: 2,
                                                col: 13,
                                            },
                                            Field(
                                                Spanned(
                                                    Location {
                                                        offset: 64,
                                                        line: 2,
                                                        col: 13,
                                                    },
                                                    FieldDef(
                                                        Spanned(
                                                            Location {
                                                                offset: 64,
                                                                line: 2,
                                                                col: 13,
                                                            },
                                                            Private,
                                                            Location {
                                                                offset: 64,
                                                                line: 2,
                                                                col: 13,
                                                            },
                                                        ),
                                                        Spanned(
                                                            Location {
                                                                offset: 64,
                                                                line: 2,
                                                                col: 13,
                                                            },
                                                            Ident(
                                                                "f",
                                                                None,
                                                            ),
                                                            Location {
                                                                offset: 65,
                                                                line: 2,
                                                                col: 14,
                                                            },
                                                        ),
                                                        Spanned(
                                                            Location {
                                                                offset: 67,
                                                                line: 2,
                                                                col: 16,
                                                            },
                                                            Ident(
                                                                "Future",
                                                                Some(
                                                                    [
                                                                        Spanned(
                                                                            Location {
                                                                                offset: 74,
                                                                                line: 2,
                                                                                col: 23,
                                                                            },
                                                                            Ident(
                                                                                "T",
                                                                                None,
                                                                            ),
                                                                            Location {
                                                                                offset: 75,
                                                                                line: 2,
                                                                                col: 24,
                                                                            },
                                                                        ),
                                                                    ],
                                                                ),
                                                            ),
                                                            Location {
                                                                offset: 76,
                                                                line: 2,
                                                                col: 25,
                                                            },
                                                        ),
                                                    ),
                                                    Location {
                                                        offset: 76,
                                                        line: 2,
                                                        col: 25,
                                                    },
                                                ),
                                            ),
                                            Location {
                                                offset: 76,
                                                line: 2,
                                                col: 25,
                                            },
                                        ),
                                    ],
                                    ret: Some(
                                        Spanned(
                                            Location {
                                                offset: 84,
                                                line: 2,
                                                col: 33,
                                            },
                                            Ident(
                                                "T",
                                                None,
                                            ),
                                            Location {
                                                offset: 85,
                                                line: 2,
                                                col: 34,
                                            },
                                        ),
                                    ),
                                    effects: [],
                                },
                                Location {
                                    offset: 85,
                                    line: 2,
                                    col: 34,
                                },
                            ),
                            Spanned(
                                Location {
                                    offset: 90,
                                    line: 3,
                                    col: 4,
                                },
                                Prototype {
                                    name: Spanned(
                                        Location {
                                            offset: 90,
                                            line: 3,
                                            col: 4,
                                        },
                                        Ident(
                                            "exec",
                                            None,
                                        ),
                                        Location {
                                            offset: 94,
                                            line: 3,
                                            col: 8,
                                        },
                                    ),
                                    args: [
                                        Spanned(
                                            Location {
                                                offset: 95,
                                                line: 3,
                                                col: 9,
                                            },
                                            Field(
                                                Spanned(
                                                    Location {
                                                        offset: 95,
                                                        line: 3,
                                                        col: 9,
                                                    },
                                                    FieldDef(
                                                        Spanned(
                                                            Location {
                                                                offset: 95,
                                                                line: 3,
                                                                col: 9,
                                                            },
                                                            Private,
                                                            Location {
                                                                offset: 95,
                                                                line: 3,
                                                                col: 9,
                                                            },
                                                        ),
                                                        Spanned(
                                                            Location {
                                                                offset: 95,
                                                                line: 3,
                                                                col: 9,
                                                            },
                                                            Ident(
                                                                "arg0",
                                                                None,
                                                            ),
                                                            Location {
                                                                offset: 99,
                                                                line: 3,
                                                                col: 13,
                                                            },
                                                        ),
                                                        Spanned(
                                                            Location {
                                                                offset: 101,
                                                                line: 3,
                                                                col: 15,
                                                            },
                                                            Ident(
                                                                "string",
                                                                None,
                                                            ),
                                                            Location {
                                                                offset: 107,
                                                                line: 3,
                                                                col: 21,
                                                            },
                                                        ),
                                                    ),
                                                    Location {
                                                        offset: 107,
                                                        line: 3,
                                                        col: 21,
                                                    },
                                                ),
                                            ),
                                            Location {
                                                offset: 107,
                                                line: 3,
                                                col: 21,
                                            },
                                        ),
                                        Spanned(
                                            Location {
                                                offset: 109,
                                                line: 3,
                                                col: 23,
                                            },
                                            Field(
                                                Spanned(
                                                    Location {
                                                        offset: 109,
                                                        line: 3,
                                                        col: 23,
                                                    },
                                                    FieldDef(
                                                        Spanned(
                                                            Location {
                                                                offset: 109,
                                                                line: 3,
                                                                col: 23,
                                                            },
                                                            Private,
                                                            Location {
                                                                offset: 109,
                                                                line: 3,
                                                                col: 23,
                                                            },
                                                        ),
                                                        Spanned(
                                                            Location {
                                                                offset: 109,
                                                                line: 3,
                                                                col: 23,
                                                            },
                                                            Ident(
                                                                "args",
                                                                None,
                                                            ),
                                                            Location {
                                                                offset: 113,
                                                                line: 3,
                                                                col: 27,
                                                            },
                                                        ),
                                                        Spanned(
                                                            Location {
                                                                offset: 115,
                                                                line: 3,
                                                                col: 29,
                                                            },
                                                            Ident(
                                                                "stringvec",
                                                                None,
                                                            ),
                                                            Location {
                                                                offset: 124,
                                                                line: 3,
                                                                col: 38,
                                                            },
                                                        ),
                                                    ),
                                                    Location {
                                                        offset: 124,
                                                        line: 3,
                                                        col: 38,
                                                    },
                                                ),
                                            ),
                                            Location {
                                                offset: 124,
                                                line: 3,
                                                col: 38,
                                            },
                                        ),
                                    ],
                                    ret: None,
                                    effects: [],
                                },
                                Location {
                                    offset: 128,
                                    line: 3,
                                    col: 42,
                                },
                            ),
                        ],
                    ),
                ),
            ),
            Location {
                offset: 130,
                line: 4,
                col: 1,
            },
        ),
    ],
)
